helix-core/src/args.rs

```diff
@@ -0,0 +1,692 @@
+
+    #[inline]
+    fn into_iter(self) -> Self::IntoIter {
+        self.positionals.iter()
+    }
+}
+
+impl<'a> AsRef<[Cow<'a, str>]> for Args<'a> {
+    #[inline]
+    fn as_ref(&self) -> &[Cow<'a, str>] {
+        self.positionals.as_ref()
+    }
+}
+
+impl PartialEq<&[&str]> for Args<'_> {
+    #[inline]
+    fn eq(&self, other: &&[&str]) -> bool {
+        let this = self.positionals.iter();
+        let other = other.iter().copied();
+
+        for (left, right) in this.zip(other) {
+            if left != right {
+                return false;
+            }
+        }
+
+        true
+    }
+}
+
+impl<'a> Index<usize> for Args<'a> {
+    type Output = str;
+
+    #[inline]
+    fn index(&self, index: usize) -> &Self::Output {
+        let cow = &self.positionals[index];
+        cow.as_ref()
+    }
+}
+
+impl<'a> Index<RangeFrom<usize>> for Args<'a> {
+    type Output = [Cow<'a, str>];
+
+    #[inline]
+    fn index(&self, index: RangeFrom<usize>) -> &Self::Output {
+        &self.positionals[index]
+    }
+}
+
+/// An iterator over an input string which yields arguments.
+///
+/// Splits on whitespace, but respects quoted substrings (using double quotes, single quotes, or backticks).
+#[derive(Debug, Clone)]
+pub struct ArgsParser<'a, const U: bool, const UB: bool> {
+    input: &'a str,
+    idx: usize,
+    start: usize,
+    unescape: bool,
+    unescape_blackslash: bool,
+}
+
+impl<'a, const U: bool, const UB: bool> ArgsParser<'a, U, UB> {
+    #[inline]
+    pub const fn is_empty(&self) -> bool {
+        self.input.is_empty()
+    }
+
+    /// Returns the args exactly as input.
+    ///
+    /// # Examples
+    /// ```
+    /// # use helix_core::args::ArgsParser;
+    /// let args = ArgsParser::from(r#"sed -n "s/test t/not /p""#);
+    /// assert_eq!(r#"sed -n "s/test t/not /p""#, args.raw());
+    ///
+    /// let args = ArgsParser::from(r#"cat "file name with space.txt""#);
+    /// assert_eq!(r#"cat "file name with space.txt""#, args.raw());
+    /// ```
+    #[inline]
+    pub const fn raw(&self) -> &str {
+        self.input
+    }
+
+    /// Returns the remainder of the args exactly as input.
+    ///
+    /// # Examples
+    /// ```
+    /// # use helix_core::args::ArgsParser;
+    /// let mut args = ArgsParser::from(r#"sed -n "s/test t/not /p""#);
+    /// assert_eq!("sed", args.next().unwrap());
+    /// assert_eq!(r#"-n "s/test t/not /p""#, args.rest());
+    /// ```
+    ///
+    /// Never calling `next` and using `rest` is functionally equivalent to calling `raw`.
+    #[inline]
+    pub fn rest(&self) -> &str {
+        &self.input[self.idx..]
+    }
+
+    /// Returns a reference to the `next()` value without advancing the iterator.
+    #[inline]
+    #[must_use]
+    #[allow(unused)]
+    fn peek(&self) -> Option<Cow<'_, str>> {
+        self.clone().next()
+    }
+}
+
+impl<'a> ArgsParser<'a, false, false> {
+    #[inline]
+    const fn parse(input: &'a str) -> ArgsParser<'a, false, false> {
```

Comment:
this isn't doing any parsing within the function call itself so let's call this `new`